compiler/rustc_mir_build/src/build/matches/test.rs

```diff
@@ -536,6 +536,7 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         test: &Test<'tcx>,
         candidate: &mut Candidate<'_, 'tcx>,
         sorted_candidates: &FxIndexMap<TestBranch<'tcx>, Vec<&mut Candidate<'_, 'tcx>>>,
+        catch_full_matched: &mut impl FnMut(&Candidate<'_, 'tcx>, &MatchPair<'_, 'tcx>),
     ) -> Option<TestBranch<'tcx>> {
```

Comment:
Please remove the callback and instead return `Option<(TestBranch, Option<Span>)>`. Then:
```rust
let ret = match ... {
  // ...
};

let matched_span = fully_matched.then_some(match_pair.pattern.span);
let ret = ret.map(|branch| (branch, matched_span));

if fully_matched {
  // ...
}
```